userd/settings.go

```diff
@@ -204,33 +200,23 @@ func (s *Settings) Set(setting, new string, sender dbus.Sender) *dbus.Error {
 	//        the X windows for _NET_WM_PID and use the windowID to
 	//        attach to zenity - not sure how this translate to the
 	//        wayland world though :/
-	txt := fmt.Sprintf(i18n.G(`
-<big><b>Allow settings change?</b></big>
-
-Allow snap %q to change %q to %q ?
-
-<span size="x-small">This dialog will close automatically after 5 minutes of inactivity.</span>
-`), snap, setting, new)
-	cmd := exec.Command("zenity", "--question", "--modal", "--text="+txt)
-	if err := cmd.Start(); err != nil {
-		return dbus.MakeFailedError(err)
-	}
-	done := make(chan error)
-	go func() { done <- cmd.Wait() }()
-	select {
-	case err = <-done:
-		// normal exit
-	case <-time.After(defaultConfirmDialogTimeout * time.Second):
-		// timeout do nothing, the other side will have timed
-		// out as well, no need to send a reply.
-		cmd.Process.Kill()
-		return nil
-	}
+	dialog, err := ui.New()
 	if err != nil {
+		return dbus.MakeFailedError(fmt.Errorf("cannot ask for settings change: %v", err))
+	}
+	answeredYes := dialog.YesNo(
+		i18n.G("Allow settings change?"),
+		fmt.Sprintf(i18n.G("Allow snap %q to change %q to %q ?"), snap, setting, new),
```

Comment:
I'm kind of -0.5 on this, let me explain why with an analogy. Imagine that you are on a train and it is approaching a station. You would hear `Next station <odd pause/><different voice>some name</different voice>`. This can almost never be translated correctly.

IMO this is a temporary solution still but perhaps we could make it a little bit closer towards the eventually integrated system settings by using an enum for the settings (or an golang interface) and then returning proper strings without %q

Not a blocker, let me know if you want me to tweak this.

Reply:
I'm probably misunderstanding (and I don't get the analogy). 

When you say `This can almost never be translated correctly.` this is a very common problem with translation I assume you mean the argument ordering?

The example from the gnu gettext manual is `printf (gettext ("String `%s' has %d characters\n"), s, strlen (s));` where for some languages the strings need to be reordered, i.e. it might be `"%d Zeichen lang ist die Zeichenkette `%s'` in German (it is a terrible translation but its just to make the point). In this case (and in similar cases with multiple %s or %q) the translated string will need to reorder things. This is possile in go with: `%[2]d Zeichen lang ist die Zeichenkette %[1]s`.

Or do you mean that the setting "default-web-browser" is meaningless to the user? Or that the desktop files are hard to read?

Reply:
@zyga
zyga 7 hours ago Member

I'm kind of -0.5 on this, let me explain why with an analogy. Imagine that you are on a train and it is approaching a station. You would hear Next station <odd pause/><different voice>some name</different voice>. This can almost never be translated correctly.

IMO this is a temporary solution still but perhaps we could make it a little bit closer towards the eventually integrated system settings by using an enum for the settings (or an golang interface) and then returning proper strings without %q

Not a blocker, let me know if you want me to tweak this.
@mvo5
mvo5 5 hours ago Member

I'm probably misunderstanding (and I don't get the analogy).

When you say This can almost never be translated correctly. this is a very common problem with translation I assume you mean the argument ordering?

The example from the gnu gettext manual is printf (gettext ("String%s' has %d characters\n"), s, strlen (s));where for some languages the strings need to be reordered, i.e. it might be"%d Zeichen lang ist die Zeichenkette %s' in German (it is a terrible translation but its just to make the point). In this case (and in similar cases with multiple %s or %q) the translated string will need to reorder things. This is possile in go with: %[2]d Zeichen lang ist die Zeichenkette %[1]s.

Or do you mean that the setting "default-web-browser" is meaningless to the user? Or that the desktop files are hard to read?
@niemeyer
The issue is not entirely clear to me as well. I think @zyga is misunderstanding what the setting value would be in this case. Note that this is a variable name.

Reply:
The issue is also not entirely clear for me. Perhaps @zyga is misunderstanding what the setting variable would mean in this case. Note that this is a variable name, so the text is okayish?

Reply:
tl;dr; i18n is hard, this is not a perfect way to do things anyway so +1

Hmm, variable names are not terribly user friendly but since this is an extended stop gap I won't block this. My analogy was that the <odd pause><different voice> indicates that those are two separate recordings stitched together and not a prerecorded sentence. One place where this will not work very well is case agreement. In some languages, polish is one of them, the case property of the variable name must agree with the rest of the sentence and this cannot be achieved with simple variable substitution.

A simple example. `Allow snap %q to change %q to %q ?` with the values of `brave` (the browser) `default browser` and `brave` (for simplicity). Let's say we have the following translation map:

`Allow snap %q to change %q to %q?` => `Czy zezwolić snapowi %q zmienić %q na %q?`
`default browser` => `domyślna przeglądarka`

Now the rendering will look like:

`Czy zezwolić snapowi "brave" zmienić "domyślna przeglądarka" na "brave"`

This looks okay except it's wrong grammatically. It has to be

`Czy zezwolić snapowi "brave" zmienić "domyślną przeglądarkę" na "brave"`

The verb "zmienić" influences the case of the subsequent nouns and other types of speech and those must be in agreement. 

I mainly brought this up because it can only be done in two ways:

1) Like a windows registry where nobody will understand what is going on as more and more things are added and it turns into a pile of goo
2) As a tailored UI with proper text and context and interactions for each element that is tunable.

Because we have limited set of settings so far those can be done as a hybrid between 1 and 2 where we have simple hacky UI but at least each message is tailored and properly worded.

But as this is a hack, it's okay to land as is. I just wanted to explain why I raised the problem.